pkg/virt-controller/watch/vmi/storage.go

```diff
@@ -265,6 +265,43 @@ func (c *Controller) updateVolumeStatus(vmi *virtv1.VirtualMachineInstance, virt
 	return nil
 }
 
+func (c *Controller) checkEphemeralHotplugVolumes(vmi *virtv1.VirtualMachineInstance, vm *virtv1.VirtualMachine) {
+	vmVolumeMap := map[string]struct{}{}
+	if vmi == nil || vm == nil {
+		return
+	}
+
+	for _, volume := range vm.Spec.Template.Spec.Volumes {
+		vmVolumeMap[volume.Name] = struct{}{}
+	}
+
+	hasEphemeral := false
+	labels := vmi.GetLabels()
+	if labels == nil {
+		labels = make(map[string]string)
+	}
+	// check if the vmi has any volumes that are not in the vm spec
+	for _, volume := range vmi.Spec.Volumes {
+		if !storagetypes.IsHotplugVolume(&volume) {
+			continue
+		}
+		if _, exists := vmVolumeMap[volume.Name]; !exists {
+			hasEphemeral = true
+			if _, ok := labels[virtv1.EphemeralHotplugLabel]; !ok {
+				// will be patched at the end of updateStatus
+				labels[virtv1.EphemeralHotplugLabel] = "true"
+				vmi.Labels = labels
+				return
+			}
+		}
+	}
+
+	// check if this vmi previously had an ephemeral volume
+	if !hasEphemeral {
```

Comment:
nit: `hasEphemeral` is not needed. Since returning early when `hasEphemeral` is true can always delete from the map here